core/src/main/java/org/sonarsource/sonarlint/core/tracking/SecurityHotspotMatchingServiceImpl.java

```diff
@@ -0,0 +1,165 @@
+
+  public SecurityHotspotMatchingServiceImpl(ConfigurationRepository configurationRepository, StorageService storageService,
+    ActiveSonarProjectBranchRepository activeSonarProjectBranchRepository, SynchronizationServiceImpl synchronizationService) {
+    this.configurationRepository = configurationRepository;
+    this.storageService = storageService;
+    this.activeSonarProjectBranchRepository = activeSonarProjectBranchRepository;
+    this.synchronizationService = synchronizationService;
+    this.executorService = Executors.newSingleThreadExecutor(r -> new Thread(r, "sonarlint-server-tracking-hotspot-updater"));
+  }
+
+  @Override
+  public CompletableFuture<MatchWithServerSecurityHotspotsResponse> matchWithServerSecurityHotspots(MatchWithServerSecurityHotspotsParams params) {
+    return CompletableFutures.computeAsync(cancelChecker -> {
+      var configurationScopeId = params.getConfigurationScopeId();
+      var effectiveBindingOpt = configurationRepository.getEffectiveBinding(configurationScopeId);
+      var activeBranchOpt = activeSonarProjectBranchRepository.getActiveSonarProjectBranch(configurationScopeId);
+      if (effectiveBindingOpt.isEmpty() || activeBranchOpt.isEmpty()) {
+        return new MatchWithServerSecurityHotspotsResponse(params.getClientTrackedHotspotsByServerRelativePath().entrySet().stream()
+          .map(e -> Map.entry(e.getKey(), e.getValue().stream()
+            .<Either<ServerMatchedSecurityHotspotDto, LocalOnlySecurityHotspotDto>>map(issue -> Either.forRight(new LocalOnlySecurityHotspotDto(UUID.randomUUID())))
+            .collect(Collectors.toList())))
+          .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue)));
+      }
+      var binding = effectiveBindingOpt.get();
+      var activeBranch = activeBranchOpt.get();
+      if (params.shouldFetchHotspotsFromServer()) {
+        refreshServerSecurityHotspots(cancelChecker, binding, activeBranch, params);
+      }
+      var clientTrackedIssuesByServerRelativePath = params.getClientTrackedHotspotsByServerRelativePath();
+      return new MatchWithServerSecurityHotspotsResponse(clientTrackedIssuesByServerRelativePath.entrySet().stream().map(e -> {
+        var serverRelativePath = e.getKey();
+        var serverHotspots = storageService.binding(binding).findings().loadHotspots(activeBranch, serverRelativePath);
+        var clientHotspotTrackables = toTrackables(e.getValue());
+        var matches = matchSecurityHotspots(serverHotspots, clientHotspotTrackables)
+          .stream().<Either<ServerMatchedSecurityHotspotDto, LocalOnlySecurityHotspotDto>>map(result -> {
+            if (result.isLeft()) {
+              var serverSecurityHotspot = result.getLeft();
+              var creationDate = serverSecurityHotspot.getCreationDate();
+              var isOnNewCode = storageService.binding(binding).newCodeDefinition().read().map(definition -> definition.isOnNewCode(creationDate)).orElse(true);
```

Comment:
We can maybe extract the read of the new code definition, to avoid doing it for each finding?

Reply:
Yep, good idea!